README.md
chi-elections
=============

[![Build Status](https://travis-ci.org/datamade/python-chicago-elections.svg?branch=master)](https://travis-ci.org/datamade/python-chicago-elections)

chi-elections is a Python package for loading and parsing election results from the [Chicago Board of Elections](http://www.chicagoelections.com/).

Summary Results
---------------

The Board of Elections provides election-night results at a racewide level.  The file lives at

http://www.chicagoelections.com/results/ap/summary.txt

before election day, for testing.

It lives at

http://www.chicagoelections.com/ap/summary.txt

on election night.

### Text layout

From http://www.chicagoelections.com/results/ap/text_layout.txt:

```
Summary Export File Format      Length  Column Position
Contest Code                    4       1-4
Candidate Number                3       5-7
Num. Eligible Precincts         4       8-11
Votes                           7       12-18
Num. Completed precincts        4       19-22
Party Abbreviation              3       23-25
Political Subdivision Abbrev    7       26-32
Contest name                    56      33-88
Candidate Name                  38      89-126
Political subdivision name      25      127-151
Vote For                        3       152-154
```

### Gotchas

Prior to election night, the test file will include all fields.  At some point on election night, the file will only contain the numeric values in the first 22 columns.

This means that you need to:

* Make sure you save candidate names in some way, like a database, before election night
* Make sure you store ballot order (Candidate Number in the text layout above) with the candidate.  You'll need to use this, in combination with Contest Code, to look up the cached candidates.

At some point at the end of election night, the results file will no longer be available at http://www.chicagoelections.com/ap/summary.txt and will be available at http://www.chicagoelections.com/results/ap/summary.txt
.  However, it will not be updated. You'll need to scrape, enter or load results in some other way if you need updates after election night.


### Results client

To access the results:

```python
from chi_elections import SummaryClient

client = SummaryClient()
client.fetch()
mayor = next(r for r in client.races if r.name == "Mayor")
self.assertEqual(len(mayor.candidates), 5)

rahm = next(c for c in mayor.candidates
            if c.full_name == "RAHM EMANUEL")
print(rahm.vote_total)
```

If you want to specify an alternate url, for example the test URL, pass it to the constructor of `SummaryClient`:

```python
client = SummaryClient(url='http://www.chicagoelections.com/results/ap/summary.txt')
```

Precinct Results
----------------

After election night, precinct-level results are published to https://chicagoelections.com/en/election-results.html.  The results are HTML files, so we have to scrape the results from HTML tables.

### Results client

To access the results:

from chi_elections import elections

```python
muni_elections = [election for name, election in
                  elections().items() if 'municipal' in name.lower()]

for election in muni_elections:
    for name, race in election.races.items():
        if 'alderman' in name.lower() or 'mayor' in name.lower():
            for precinct, votes in race.precincts.items():
                print(precinct, votes)
```                

### Command Line Interface

To download a CSV version of the summary file, run:

    chi_elections summary > results.csv

To hit the test file:

    chi_elections summary --test > results.csv



CHANGELOG.md
CHANGELOG
=========

0.3
---

* Python 3 only
* Refactored precinct results for updated CBOE website.
* Added CI
* Removed precinct CLI (might bring this back later)

0.2
---

* Handle utf-8 characters
* Replace two single quotes in candidate names with a single double quote
* Add command line interface to output CSV of results


chi_elections/cli.py
import codecs
import sys

import csv

import click

from .constants import SUMMARY_URL, TEST_SUMMARY_URL
from .precincts import Election
from .summary import SummaryClient, SummaryParser

if sys.version_info < 3:
    # Wrap sys.stdout into a StreamWriter to allow writing unicode.
    # See https://wiki.python.org/moin/PrintFails
    sys.stdout = codecs.getwriter('utf-8')(sys.stdout)

@click.group()
def main():
    pass

@click.command()
@click.option('-f', '--file', type=click.File())
@click.option('--test/--no-test', default=False)
def summary(file, test):
    if test:
        url = TEST_SUMMARY_URL
    else:
        url = SUMMARY_URL

    if file:
        parser = SummaryParser()
        parser.parse(file.read())
        races = parser.races
    else:
        client = SummaryClient(url=url)
        client.fetch()
        races = client.races

    fieldnames = [
       'contest_code',
       'race_name',
       'precincts_total',
       'precincts_reporting',
       'vote_for',
       'candidate_number',
       'full_name',
       'party',
       'vote_total',
    ]
    writer = csv.DictWriter(sys.stdout, fieldnames=fieldnames)
    writer.writeheader()
    for race in races:
        race_attrs = race.serialize()
        for candidate_result in race.candidates:
            row = dict(**race_attrs)
            row.update(candidate_result.serialize())
            writer.writerow(row)

main.add_command(summary)


chi_elections/precincts.py
"""
Parse tabular precinct-level results.
"""
import functools
import collections

import lxml.html
import requests

class Election(object):
    ELECTION_URL = 'https://chicagoelections.com/en/election-results.asp'
    
    def __init__(self, elec_code, name, session):
        self.elec_code = elec_code
        self.name = name
        self.url = self.ELECTION_URL
        self.session = session
        self._turnout = None

    @property
    def turnout(self):
        if not self._turnout:
            self.races
        return self._turnout

    @property
    @functools.lru_cache(maxsize=1)
    def races(self):
        response = self.session.get(self.url,
                                    params={'election': self.elec_code})
        page = lxml.html.fromstring(response.text)
        option_els = page.xpath(
            "//select[@name='race']/option")

        races = {}

        for option_el in option_els:
            name = option_el.text
            race_num = option_el.get('value')
            if 'registered voters' in name.lower():
                self._turnout = Race(self.elec_code,
                                     name, race_num, self.session)
            else:
                races[option_el.text] = Race(self.elec_code,
                                             name,
                                             race_num, self.session)

        return races
    
class Race(object):
    RESULTS_URL = 'https://chicagoelections.com/en/data-export.asp'
    
    def __init__(self, elec_code, name=None, number=None, session=None):
        self.elec_code = elec_code
        self.number = number
        self.name = name
        self.session = session

    def __str__(self):
        if self.name and self.number:
            return "{} ({})".format(self.name, self.number)
        elif self.name:
            return self.name
        elif self.number:
            return self.number
        else:
            return ""

    @property
    @functools.lru_cache(maxsize=1)
    def precincts(self):
        results_d = {}
        
        response = self.session.get(self.RESULTS_URL,
                                    params = {'election': self.elec_code,
                                              'race': self.number})
        page = lxml.html.fromstring(response.text)

        tables = page.xpath('//table')

        total = tables.pop(0)
        header_row = total.xpath('./tr')[0]
        keys = ['precinct']
        for cell in header_row.xpath('./td//text()'):
            keys.append(cell.strip().replace("''", "'"))

        for table in tables:
            rows = table.xpath('./tr')
            ward = rows.pop(0).xpath('./td//text()')[0]
            ward_num = int(ward.split()[-1])
            rows.pop(0) # ignore repeated header row
            for row in rows:
                votes = {}
                for key, cell in zip(keys, row.xpath('./td//text()')):
                    if cell == 'Total': # ignore ward subtotals
                        break
                    if '%' in key or key == 'Votes': # ignore derived vars
                        continue
                    votes[key] = int(cell.strip().replace(',', ''))
                else:
                    precinct = votes.pop('precinct')
                    results_d[(ward_num, precinct)] = votes

        return results_d

    @property
    @functools.lru_cache(maxsize=1)
    def wards(self):
        results_d = collections.defaultdict(lambda: collections.defaultdict(int))
        for (ward, precinct), precinct_votes in self.precincts.items():
            for choice, votes in precinct_votes.items():
                results_d[ward][choice] += votes

        results_d = dict(results_d)
        for k, v in results_d.items():
            results_d[k] = dict(v)

        return results_d

    @property
    @functools.lru_cache(maxsize=1)
    def total(self):
        results_d = collections.defaultdict(int)
        for precinct_votes in self.precincts.values():
            for choice, votes in precinct_votes.items():
                results_d[choice] += votes

        return dict(results_d)

            

def elections(session=None):
    '''List all available elections'''

    election_url = 'https://chicagoelections.com/en/election-results.html'

    if session is None:
        session = requests.Session()
    else:
        session = session
    
    response = session.get(election_url)
    page = lxml.html.fromstring(response.text)

    election_links = page.xpath("//a[starts-with(@href, 'election-results.asp?election=')]")

    elex = {}

    for link in election_links:
        name = link.text
        election_code = link.get('href').split('=')[-1]
        elex[name] = Election(election_code, name, session)

    return elex
        


chi_elections/summary.py
"""
Parse fixed-width summary file.

This file lives at

http://www.chicagoelections.com/results/ap/

before election day, for testing.

It lives at

www.chicagoelections.com/ap/

on election night.

This file provides racewide results.

"""
from collections import OrderedDict

import requests

from .constants import SUMMARY_URL
from .transforms import replace_single_quotes

class FixedWidthField(object):
    def __init__(self, index, length, transform=None):
        self.index = index
        self.length = length
        self.transform = transform
        self.name = None

    def parse(self, s):

        val = s[self.index:self.index + self.length]
        val = val.strip()
        if self.transform is None:
            return val
        else:
            try:
                return self.transform(val)
            except ValueError:
                return None


class FixedWidthParserMeta(type):
    def __new__(cls, name, parents, dct):
        dct['_fields'] = []
        for k, v in list(dct.items()):
            if isinstance(v, FixedWidthField):
                v.name = k
                dct['_fields'].append(v)
                del dct[k]

        new_cls = super(FixedWidthParserMeta, cls).__new__(cls, name, parents, dct)
        return new_cls


class FixedWidthParser(object, metaclass=FixedWidthParserMeta):

    def parse_line(self, line):
        attrs = {}
        for field in self._fields:
            attrs[field.name] = field.parse(line)

        return attrs 


class ResultParser(FixedWidthParser):
    # Summary Export File Format           Length    Column Position
    # Contest Code                         4         1-4
    # Candidate Number                     3         5-7
    # # of Eligible Precincts              4         8-11
    # Votes                                7         12-18
    # # Completed precincts                4         19-22
    # Party Abbreviation                   3         23-25
    # Political Subdivision Abbreviation   7         26-32
    # Contest name                         56        33-88
    # Candidate Name                       38        89-126
    # Political subdivision name           25        127-151
    # Vote For                             3         152-154
    contest_code = FixedWidthField(0, 4, transform=int)
    candidate_number = FixedWidthField(4, 3, transform=int)
    precincts_total = FixedWidthField(7, 4, transform=int)
    vote_total = FixedWidthField(11, 7, transform=int)
    precincts_reporting = FixedWidthField(18, 4, transform=int)
    party = FixedWidthField(22, 3)
    reporting_unit_name = FixedWidthField(25, 7)
    race_name = FixedWidthField(32, 56)
    candidate_name = FixedWidthField(88, 38, transform=replace_single_quotes)
    reporting_unit_name = FixedWidthField(126, 25)
    vote_for = FixedWidthField(151, 3, transform=int)


class Result(object):
    def __init__(self, candidate_number, full_name, party, race, vote_total,
            reporting_unit_name):
        self.candidate_number = candidate_number
        self.full_name = full_name
        self.party = party
        self.race = race
        self.vote_total = vote_total

    def __str__(self):
        return "{}: {}d".format(self.name, self.vote_total)

    def serialize(self):
        return OrderedDict((
            ('candidate_number', self.candidate_number),
            ('full_name', self.full_name),
            ('party', self.party),
            ('vote_total',self.vote_total),
        ))


class Race(object):
    def __init__(self, contest_code, name, precincts_total=0,
            precincts_reporting=0, vote_for=1):
        self.contest_code = contest_code
        self.name = name
        self.candidates = []
        self.precincts_total = precincts_total
        self.precincts_reporting = precincts_reporting
        self.vote_for = vote_for

    def serialize(self):
        return OrderedDict((
            ('contest_code', self.contest_code),
            ('race_name', self.name),
            ('precincts_total', self.precincts_total),
            ('precincts_reporting', self.precincts_reporting),
            ('vote_for', self.vote_for),
        ))

    def __str__(self):
        return self.name


class SummaryParser(object):
    def __init__(self):
        self._result_parser = ResultParser()

    def parse(self, s):
        self.races = []
        self._race_lookup = {}

        for line in s.splitlines(True):
            parsed = self._result_parser.parse_line(line)
            race = self.get_or_create_race(parsed)
            result = Result(
                candidate_number=parsed['candidate_number'],    
                vote_total=parsed['vote_total'],
                party=parsed['party'],
                race=race,
                full_name=parsed['candidate_name'],
                reporting_unit_name=parsed['reporting_unit_name'],
            )
            race.candidates.append(result)
    
    def get_or_create_race(self, attrs):
        try:
            race = self._race_lookup[attrs['contest_code']]
        except KeyError:
            race = Race(
                contest_code=attrs['contest_code'],
                name=attrs['race_name'],
                precincts_total=attrs['precincts_total'],
                precincts_reporting=attrs['precincts_reporting'],
                vote_for=attrs['vote_for'],
            )
            self._race_lookup[attrs['contest_code']] = race
            self.races.append(race)
        
        return race


class SummaryClient(object):
    DEFAULT_URL = SUMMARY_URL 

    def __init__(self, url=None):
        if url is None:
            url = self.DEFAULT_URL
        self._url = url

        self._parser = SummaryParser()

    def get_url(self):
        return self._url

    def fetch(self):
        url = self.get_url()
        r = requests.get(url)
        self._parser.parse(r.text)

    @property
    def races(self):
        return self._parser.races
